contracts/coconsensus/Coconsensus.sol

```diff
@@ -46,23 +47,310 @@ contract Coconsensus is MasterCopyNonUpgradable, GenesisCoconsensus, MosaicVersi
+             * skip the protocore setup if the current metachain id is equal to
+             * the auxiliary metachain id.
+             */
+            if (currentMetachainId != auxiliaryMetachainId) {
+                // Setup protocore contract for the given metachain id.
+                setupProtocores(currentMetachainId);
+            }
+
+            // Setup observer contract for the given metachain id.
+            setupObservers(currentMetachainId);
+
+            // Traverse to next metachain id from the link list mapping.
+            currentMetachainId = genesisMetachainIds[currentMetachainId];
+        }
+    }
+
+
+    /* External Functions */
+
+    /**
+     * @notice finaliseCheckpoint() function finalises a checkpoint at
+     *         a metachain.
+     *
+     * @param _metachainId A metachain id to finalise a checkpoint.
+     * @param _blockNumber A block number of a checkpoint.
+     * @param _blockHash A block hash of a checkpoint.
+     *
+     * \pre `_metachainId` is not 0.
+     * \pre `_blockHash` is not 0.
+     * \pre `msg.sender` should be the protocore contract.
+     * \pre `_blockNumber` must be multiple of epoch length.
+     * \pre `_blockNumber` must be greater than the last finalized block number.
+     *
+     * \post Sets the `blockchains` mapping.
+     * \post Sets the `blockTips` mapping.
+     */
+    function finaliseCheckpoint(
+        bytes32 _metachainId,
+        uint256 _blockNumber,
+        bytes32 _blockHash
+    )
+        external
+        onlyRunningProtocore(_metachainId)
+    {
+        // Check if the metachain id is not null.
+        require(
+            _metachainId != bytes32(0),
+            "Metachain id must not be null."
+        );
+
+        // Check if the blockhash is not null.
+        require(
+            _blockHash != bytes32(0),
+            "Blockhash must not be null."
+        );
+
+        /*
+         * Check if the new block number is greater than the last
+         * finalised block number.
+         */
+        uint256 lastFinalisedBlockNumber = blockTips[_metachainId];
+        require(
+            _blockNumber > lastFinalisedBlockNumber,
+            "The block number of the checkpoint must be greater than the block number of last finalised checkpoint."
+        );
+
+        // Check if the block number is multiple of epoch length.
+        ProtocoreI protocore = protocores[_metachainId];
+        uint256 epochLength = protocore.epochLength();
+        require(
+            (_blockNumber % epochLength) == 0,
+            "Block number must be a checkpoint."
+        );
+
+        // Store the finalised block in the mapping.
+        Block storage finalisedBlock = blockchains[_metachainId][_blockNumber];
+        finalisedBlock.blockHash = _blockHash;
+        finalisedBlock.commitStatus = CheckpointCommitStatus.Finalized;
+
+        ProtocoreI selfProtocore = protocores[auxiliaryMetachainId];
```

Comment:
This part should be updated as it has been agreed to keep `relative dynasty number` in coconsensus.

Reply:
Now using `relativeSelfDynasty`